apis/apiextensions/v1/composition_revision_types.go

```diff
@@ -41,33 +41,64 @@ type CompositionRevisionSpec struct {
 	// +immutable
 	CompositeTypeRef TypeReference `json:"compositeTypeRef"`
 
-	// PatchSets define a named set of patches that may be included by
-	// any resource in this Composition.
-	// PatchSets cannot themselves refer to other PatchSets.
+	// Mode controls what type or "mode" of Composition will be used.
+	//
+	// "Resources" (the default) indicates that a Composition uses what is
+	// commonly referred to as "Patch & Transform" or P&T composition. This mode
+	// of Composition uses an array of resources, each a template for a composed
+	// resource.
+	//
+	// "Pipeline" indicates that a Composition specifies a pipeline
+	// of Composition Functions, each of which is responsible for producing
+	// composed resources that Crossplane should create or update. THE PIPELINE
+	// MODE IS A BETA FEATURE. It is not honored if the relevant Crossplane
+	// feature flag is disabled.
+	// +optional
+	// +kubebuilder:validation:Enum=Resources;Pipeline
+	// +kubebuilder:default=Resources
+	Mode *CompositionMode `json:"mode,omitempty"`
+
+	// PatchSets define a named set of patches that may be included by any
+	// resource in this Composition. PatchSets cannot themselves refer to other
+	// PatchSets.
+	//
+	// PatchSets are only used by the "Resources" mode of Composition. They
+	// are ignored by other modes.
```

Comment:
I guess it's too late, but it starts to feel like we could have implemented composition through functions in a dedicated resource altogether.

Reply:
We did consider that back in the day. Using `Composition` feels right to me though. It's doing the same task as P&T Composition, we already have all the machinery to reference and select Compositions, etc.